### PR TITLE
Running C tests after build

### PR DESCRIPTION
**Pull request by jchen351**
This pull request includes changes to the GitHub Actions workflows and the `test/CMakeLists.txt` file. The changes are aimed at improving the build and test process for different platforms and configurations. The changes include updating the `actions/checkout` action to version 4 and enabling submodules, removing the redundant `git submodule update` steps, restructuring the Docker-based build and test steps, adding test execution steps, and reorganizing the `unit_tests` executable in the `test/CMakeLists.txt` file.

Changes to GitHub Actions workflows:

* `.github/workflows/linux-cpu-arm64-build.yml`, `.github/workflows/linux-cpu-x64-build.yml`, `.github/workflows/linux-gpu-x64-build.yml`, `.github/workflows/win-cpu-arm64-build.yml`, `.github/workflows/win-cpu-x64-build.yml`, `.github/workflows/win-gpu-x64-build.yml`: Updated `actions/checkout` to version 4 and enabled submodules. Removed the redundant `git submodule update` steps. [[1]](diffhunk://#diff-41c373598f771a9d90102d4d380ba704651d296b865d6974fe3bbd7d019b35c1L15-R17) [[2]](diffhunk://#diff-e758ed36b14737d7c89353329cb735a6d081f5fa288b809d49e47bfc0ade96c3L15-R17) [[3]](diffhunk://#diff-e941545683ca21b146fc6f414d9fe9e311f5f5efc44fce47287e8b0d835434aeL17-R19) [[4]](diffhunk://#diff-df45fc531e16b591e484e29a3a4c836e876f3c8136137ffa39bb6979af6b1c79L22-R26) [[5]](diffhunk://#diff-14c154d7bf4d7c90f647475b0340d0840319223d91faa598c4756739555bd863L25-R28) [[6]](diffhunk://#diff-0989dea22643962b5f5d6f36719798b1975f96bf0537ff0053c57c5e9a774e95L21-R25)

* `.github/workflows/linux-cpu-arm64-build.yml`, `.github/workflows/linux-gpu-x64-build.yml`: Restructured the Docker-based build and test steps to separate the configuration, build, and test execution steps. [[1]](diffhunk://#diff-41c373598f771a9d90102d4d380ba704651d296b865d6974fe3bbd7d019b35c1L32-R63) [[2]](diffhunk://#diff-e941545683ca21b146fc6f414d9fe9e311f5f5efc44fce47287e8b0d835434aeL50-R83)

* `.github/workflows/linux-cpu-x64-build.yml`, `.github/workflows/win-cpu-arm64-build.yml`, `.github/workflows/win-cpu-x64-build.yml`, `.github/workflows/win-gpu-x64-build.yml`: Added steps to execute unit tests. [[1]](diffhunk://#diff-e758ed36b14737d7c89353329cb735a6d081f5fa288b809d49e47bfc0ade96c3L32-R43) [[2]](diffhunk://#diff-df45fc531e16b591e484e29a3a4c836e876f3c8136137ffa39bb6979af6b1c79R56-R60) [[3]](diffhunk://#diff-14c154d7bf4d7c90f647475b0340d0840319223d91faa598c4756739555bd863R63-R67) [[4]](diffhunk://#diff-0989dea22643962b5f5d6f36719798b1975f96bf0537ff0053c57c5e9a774e95R58-R62)

Changes to `test/CMakeLists.txt`:

* [`test/CMakeLists.txt`](diffhunk://#diff-33394812ba204689144fd2f80832db83853ba1cb32403edb4e15fe4893e675fdL9-R15): Reorganized the `unit_tests` executable to include `c_api_tests.cpp`, `model_tests.cpp`, `sampling_tests.cpp`, and `sampling_benchmark.cpp`.

### CODE REVIEW

File: test/test_models/llama/config.json
Comment (jchen351) [inline test/test_models/llama/config.json:1]: These model, llama and phi-2, were copied from examples.

File: test/test_models/phi-2/config.json
Comment (jchen351) [inline test/test_models/phi-2/config.json:1]: This and phi-2, were copied from examples/ folder

File: tools/ci_build/github/linux/docker/inference/x64/default/gpu/Dockerfile [hunk 1]
```diff
@@ -2,15 +2,17 @@
 # Licensed under the MIT License.
 
 # This file is used by Zip-Nuget Packaging NoContribOps Pipeline,Zip-Nuget-Java Packaging Pipeline
-ARG BASEIMAGE=nvidia/cuda:12.2.2-cudnn8-devel-ubi8
+ARG BASEIMAGE=nvidia/cuda:11.8.0-cudnn8-devel-ubi8
 FROM $BASEIMAGE
 
-ENV PATH="/usr/local/cuda/binet:/opt/rh/gcc-toolset-12/root/usr/bin:${PATH}"
+ENV PATH="/usr/local/nvidia/bin:/usr/local/cuda/bin:/opt/rh/gcc-toolset-12/root/usr/bin:${PATH}"
```
Comment (snnn) [inline tools/ci_build/github/linux/docker/inference/x64/default/gpu/Dockerfile:8]: What's in the /usr/local/nvidia/bin folder?

File: tools/ci_build/github/linux/docker/inference/x64/default/gpu/Dockerfile [hunk 2]
```diff
@@ -2,15 +2,17 @@
...
 
-ENV PATH="/usr/local/cuda/binet:/opt/rh/gcc-toolset-12/root/usr/bin:${PATH}"
+ENV PATH="/usr/local/nvidia/bin:/usr/local/cuda/bin:/opt/rh/gcc-toolset-12/root/usr/bin:${PATH}"
 ENV LANG=en_US.UTF-8
 ENV LC_ALL=en_US.UTF-8
-ENV CUDA_HOME=/usr/local/cuda
-ENV CUDNN_HOME=/usr/lib/x86_64-linux-gnu/
-ENV LD_LIBRARY_PATH="/usr/local/cuda/lib64:${LD_LIBRARY_PATH}"
+ENV CUDA_HOME=/usr/local/cuda-11.8
+ENV CUDA_PATH=/usr/local/cuda-11.8
```
Comment (snnn) [inline tools/ci_build/github/linux/docker/inference/x64/default/gpu/Dockerfile:12]: I remember CUDA 11.8 is not compatible with GCC 12. DId you get it work?

File: .github/workflows/linux-cpu-x64-build.yml [hunk 3]
```diff
@@ -29,23 +32,12 @@ jobs:
...
-          cmake --build --preset linux_gcc_cpu_release
-          
-      - name: Build with CMake and clang
-        run: |
-          set -e -x
-          rm -rf build
-          cmake --preset linux_clang_cpu_release
-          cmake --build --preset linux_clang_cpu_release
+          cmake --preset linux_${{ matrix.compiler }}_cpu_release
+          cmake --build --preset linux_${{ matrix.compiler }}_cpu_release
```
Comment (snnn) [inline .github/workflows/linux-cpu-x64-build.yml:40]: Why the arm one use docker but this one doesn't?

File: .github/workflows/linux-gpu-x64-build.yml [hunk 4]
```diff
@@ -9,16 +9,28 @@ env:
...
+  linux-gpu-x64-build:
     runs-on: [ "self-hosted", "1ES.Pool=onnxruntime-genai-Ubuntu2004-T4" ]
     steps:
       - name: Checkout OnnxRuntime GenAI repo
-        uses: actions/checkout@v2
-
+        uses: actions/checkout@v4
+        with:
+          submodules: true
+#      - name: Install Nvidia Container Toolkit
```
Comment (snnn) [inline .github/workflows/linux-gpu-x64-build.yml:23]: If they are not needed ,remove them.

File: .github/workflows/linux-cpu-x64-build.yml (hunk 3 above)
Comment (jchen351) [reply .github/workflows/linux-cpu-x64-build.yml:40]: It is fast to run natively

Comment (jchen351) [reply .github/workflows/linux-cpu-x64-build.yml:40]: If you prefer, we make them all run with Manylinux after pre-release, consistent with packingng pipeline. This way, we might be able to reuse the artifact, and consolidate Dockerfiles.

**Pull request merged**